### PR TITLE
[rllib] Part 2 of multiagent support

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

This implements batched multi-agent execution, adds support to a few of the algorithms and optimizers, and some basic tests and examples.

- [x] Wait for https://github.com/ray-project/ray/pull/2268 to merge
## Related issue number

https://github.com/ray-project/ray/issues/2053

### CODE REVIEW

File: python/ray/rllib/utils/tf_run_builder.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        base_index = len(self.fetches)
+        self.fetches.extend(fetches)
+        return list(range(base_index, len(self.fetches)))
+
+    def get(self, to_fetch):
+        if self._executed is None:
+            try:
+                self._executed = run_timeline(
+                    self.session, self.fetches, self.debug_name,
+                    self.feed_dict, os.environ.get("TF_TIMELINE_DIR"))
```
Comment (richardliaw) [inline python/ray/rllib/utils/tf_run_builder.py]: should document this

File: python/ray/rllib/utils/sampler.py [hunk 2]
```diff
@@ -209,11 +218,13 @@ def get_batch_builder():
                     dict(episode.agent_rewards))
             else:
                 all_done = False
+                # At least send an empty dict if not done
+                actions_to_send[env_id]
```
Comment (richardliaw) [inline python/ray/rllib/utils/sampler.py]: this line is just sitting here ..

File: python/ray/rllib/optimizers/sample_batch.py [hunk 3]
```diff
@@ -107,7 +111,7 @@ def postprocess_batch_so_far(self):
                 pre_batch, other_batches)
 
         # Append into policy batches and reset
-        for agent_id, post_batch in post_batches.items():
+        for agent_id, post_batch in sorted(post_batches.items()):
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/sample_batch.py]: why is sorting needed?

**Review by richardliaw — APPROVED** [review]
works fine

File: python/ray/rllib/optimizers/sample_batch.py (hunk 3 above)
Comment (ericl) [reply python/ray/rllib/optimizers/sample_batch.py]: It keeps them in a deterministic order for tests.

File: python/ray/rllib/utils/sampler.py (hunk 2 above)
Comment (ericl) [reply python/ray/rllib/utils/sampler.py]: This creates an empty entry in the dict

Comment (ericl) [reply python/ray/rllib/utils/sampler.py]: I'll add an assign...b

**Pull request merged**